main.c

```diff
@@ -524,6 +528,139 @@ init_vt(struct vkcube *vc)
+          .sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO,
+          .pNext = &mod_create_info,
+          .imageType = VK_IMAGE_TYPE_2D,
+          .format = vc->image_format,
+          .extent = (VkExtent3D){
+              gbm_bo_get_width(b->gbm_bo),
+              gbm_bo_get_height(b->gbm_bo),
+              1
+          },
+          .mipLevels = 1,
+          .arrayLayers = 1,
+          .samples = 1,
+          .tiling = VK_IMAGE_TILING_DRM_FORMAT_MODIFIER_EXT,
+          .usage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT,
+          .initialLayout = VK_IMAGE_LAYOUT_UNDEFINED,
+      };
+
+      res = vkCreateImage(vc->device, &base_create_info,
+                          NULL, &b->image);
+      fail_if(res != VK_SUCCESS, "failed to create image");
+
+      PFN_vkGetMemoryFdPropertiesKHR getMemoryFdPropertiesKHR =
+          (void *)vkGetDeviceProcAddr(vc->device, "vkGetMemoryFdPropertiesKHR");
+      fail_if(getMemoryFdPropertiesKHR == NULL,
+              "vkGetDeviceProcAddr(\"vkGetMemoryFdPropertiesKHR\") failed");
+
+      int prime_fd = gbm_bo_get_fd(b->gbm_bo);
+      fail_if(prime_fd < 0, "failed to get prime fd for gbm_bo");
+
+      VkMemoryFdPropertiesKHR fd_props = {
+          .sType = VK_STRUCTURE_TYPE_MEMORY_FD_PROPERTIES_KHR,
+      };
+      getMemoryFdPropertiesKHR(vc->device,
+                                    VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT,
+                                    prime_fd, &fd_props);
+
+      /* get image memory requirements */
+      const VkImageMemoryRequirementsInfo2 mem_reqs_info = {
+          .sType = VK_STRUCTURE_TYPE_IMAGE_MEMORY_REQUIREMENTS_INFO_2,
+          .image = b->image,
+      };
+      VkMemoryRequirements2 mem_reqs = {
+          .sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2,
+      };
+      vkGetImageMemoryRequirements2(vc->device, &mem_reqs_info, &mem_reqs);
+
+      const uint32_t memory_type_bits = fd_props.memoryTypeBits &
+          mem_reqs.memoryRequirements.memoryTypeBits;
+      fail_if(!memory_type_bits, "no valid memory type");
+
+      const VkMemoryDedicatedAllocateInfo memory_dedicated_info = {
+          .sType = VK_STRUCTURE_TYPE_MEMORY_DEDICATED_ALLOCATE_INFO,
+          .image = b->image,
+      };
+      const VkImportMemoryFdInfoKHR memory_fd_info = {
+          .sType = VK_STRUCTURE_TYPE_IMPORT_MEMORY_FD_INFO_KHR,
+          .pNext = &memory_dedicated_info,
+          .handleType = VK_EXTERNAL_MEMORY_HANDLE_TYPE_DMA_BUF_BIT_EXT,
+          .fd = prime_fd,
+      };
+      res = vkAllocateMemory(vc->device,
+          &(VkMemoryAllocateInfo){
+          .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+          .pNext = &memory_fd_info,
+          .allocationSize = mem_reqs.memoryRequirements.size,
+          .memoryTypeIndex = ffs(memory_type_bits) - 1,
+          },
+          NULL, &b->mem);
+      fail_if(res != VK_SUCCESS, "vkAllocateMemory");
+
+      res = vkBindImageMemory(vc->device, b->image, b->mem, 0);
```

Comment:
Doesn't `VkMemoryDedicatedAllocateInfo` already bind `image` to `mem`?

Reply:
No, I'm getting black screen without this.

Reply:
I must have been caught by the poor wording of https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkMemoryDedicatedAllocateInfo.html, which first says:

> If the _pNext_ chain includes a _VkMemoryDedicatedAllocateInfo_ structure, then that structure includes a handle of the sole buffer or image resource that the memory **can** be bound to.

But then further goes on to say for the parameters (which is what I read to say this):

> `image` is [`VK_NULL_HANDLE`](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_NULL_HANDLE.html) or a handle of an image which this memory will be bound to.

"will be"?

Reply:
No idea. vk_glow.c, which I used as a model / source, also  calls vkBindImageMemory

Reply:
I think you need to call it, the documentation wording is simply a bit inconsistent. Let's mark this as resolved.

Reply:
agreed. Thanks for the review!